narwhal/config/src/committee.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{CommitteeUpdateError, ConfigError, Epoch, Stake};
+use crypto::{NetworkPublicKey, PublicKey};
+use fastcrypto::traits::EncodeDecodeBase64;
+use multiaddr::Multiaddr;
+use once_cell::sync::OnceCell;
+use rand::rngs::StdRng;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
+pub struct Authority {
+    /// The id under which we identify this authority across Narwhal
+    #[serde(skip)]
+    pub id: OnceCell<AuthorityIdentifier>,
+    /// The authority's main PublicKey which is used to verify the content they sign.
+    pub protocol_key: PublicKey,
+    /// The voting power of this authority.
+    pub stake: Stake,
+    /// The network address of the primary.
+    pub primary_address: Multiaddr,
+    /// Network key of the primary.
+    pub network_key: NetworkPublicKey,
+}
+
+impl Authority {
+    pub fn new(
+        protocol_key: PublicKey,
+        stake: Stake,
+        primary_address: Multiaddr,
+        network_key: NetworkPublicKey,
+    ) -> Self {
+        Self {
+            id: Default::default(),
+            protocol_key,
+            stake,
+            primary_address,
+            network_key,
+        }
+    }
+
+    pub fn id(&self) -> &AuthorityIdentifier {
+        self.id
+            .get()
+            .expect("Identifier should have been initialised by now")
+    }
+
+    pub fn protocol_key(&self) -> PublicKey {
```

Comment:
Should I guess this is called once and/or clone's complexity is insignificant (no serialization/deserialization is happening)?

Reply:
This is now returned as ref. Now, the usage is quite limited after the changes and practically no ser/des takes place - just pure look up - we can push further to get rid of them but it will require more refactoring in the `WorkerCache` which as per the PR description I've avoided doing as would increase PR's scope. Also this is used when we  need to "translate" the AuthorityIdentifier to a `AuthorityName` for SUI - we can't avoid that.